server/queue.lua

```diff
@@ -0,0 +1,239 @@
+---@class QueueType
+---@field name string
+---@field color string?
+---@field predicate (fun(source: Source): boolean)? 
+
+if GetConvar('qbx:enablequeue', 'true') == 'false' then return end
+
+-- Disable hardcap because it kicks the player when the server is full
+
+---@param resource string
+AddEventHandler('onResourceStarting', function(resource)
+    if resource == 'hardcap' then
+        lib.print.info('Preventing hardcap from starting...')
+        CancelEvent()
+    end
+end)
+
+if GetResourceState('hardcap'):find('start') then
+    lib.print.info('Stopping hardcap...')
+    StopResource('hardcap')
+end
+
+-- Queue code
+
+local config = require 'config.server'.queue
+local maxPlayers = GlobalState.MaxPlayers
+
+---@class SubQueue
+---@field size integer
+---@field positions table<string, integer>
+---@field enqueue fun(self: SubQueue, license: string)
+---@field dequeue fun(self: SubQueue, license: string)
+
+local SubQueue = {}
+local prototype = {}
+local totalSize = 0
+
+---@return SubQueue
+function SubQueue.new()
+    local self = setmetatable({}, { __index = prototype })
+    self.size = 0
+    self.positions = {}
+    return self
+end
+
+---@param self SubQueue
+---@param license string
+function prototype.enqueue(self, license)
+    self.size += 1
+    totalSize += 1
+    self.positions[license] = self.size
+end
+
+---@param self SubQueue
+---@param license string
+function prototype.dequeue(self, license)
+    local pos = self.positions[license]
+    self.positions[license] = nil
+    self.size -= 1
+    totalSize -= 1
```

Comment:
`self.size` is the size of the sub-queue of the player and `totalSize` is the total size of all queues combined. I created `totalSize` instead of a calculate function to not have to calculate the value all the time for all players every second.